gadget/validate_test.go

```diff
@@ -43,65 +43,121 @@ func (s *validateGadgetTestSuite) SetUpTest(c *C) {
-		{label: "ubuntu-save", err: `label "ubuntu-save" is reserved`},
+		// ok to allow hybrid 20-ready devices
+		{label: "ubuntu-boot"},
+		{label: "ubuntu-save"},
+		// reserved only if seed present/expected
+		{label: "ubuntu-boot", err: `label "ubuntu-boot" is reserved`, model: uc20Mod},
+		{label: "ubuntu-save", err: `label "ubuntu-save" is reserved`, model: uc20Mod},
 		// these are ok
 		{role: "system-boot", label: "ubuntu-boot"},
 		{label: "random-ubuntu-label"},
 	} {
-		err := gadget.RuleValidateVolumeStructure(&gadget.VolumeStructure{
-			Type:       "21686148-6449-6E6F-744E-656564454649",
-			Role:       tc.role,
-			Filesystem: "ext4",
-			Label:      tc.label,
-			Size:       10 * 1024,
-		})
+		gi := &gadget.Info{
+			Volumes: map[string]*gadget.Volume{
+				"vol0": {
+					Structure: []gadget.VolumeStructure{{
+						Type:       "21686148-6449-6E6F-744E-656564454649",
+						Role:       tc.role,
+						Filesystem: "ext4",
+						Label:      tc.label,
+						Size:       10 * 1024,
+					}},
+				},
+			},
+		}
+		err := gadget.Validate(gi, tc.model, nil)
 		if tc.err == "" {
 			c.Check(err, IsNil)
 		} else {
-			c.Check(err, ErrorMatches, tc.err)
+			c.Check(err, ErrorMatches, ".*: "+tc.err)
 		}
 	}
 
 }
 
-func (s *validateGadgetTestSuite) TestEnsureVolumeRuleConsistency(c *C) {
-	state := func(seed bool, label string) *gadget.ValidationState {
-		systemDataVolume := &gadget.VolumeStructure{Label: label}
-		systemSeedVolume := (*gadget.VolumeStructure)(nil)
-		if seed {
-			systemSeedVolume = &gadget.VolumeStructure{}
+// rolesYaml produces gadget metadata with volumes with structure withs the given
+// role if data, seed or save are != "-", and with their label set to the value
+func rolesYaml(c *C, data, seed, save string) *gadget.Info {
+	h := `volumes:
+  roles:
+    schema: gpt
+    bootloader: grub
+    structure:
+`
+	if data != "-" {
+		h += `
+      - name: data
+        size: 1G
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        role: system-data
+`
+		if data != "" {
+			h += fmt.Sprintf("        filesystem-label: %s\n", data)
 		}
-		return &gadget.ValidationState{
-			SystemSeed: systemSeedVolume,
-			SystemData: systemDataVolume,
+	}
+	if seed != "-" {
+		h += `
+      - name: seed
+        size: 1G
+        type: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+        role: system-seed
+`
+		if seed != "" {
+			h += fmt.Sprintf("        filesystem-label: %s\n", seed)
 		}
 	}
 
+	if save != "-" {
+		h += `
+      - name: save
+        size: 32M
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        role: system-save
+`
+		if save != "" {
+			h += fmt.Sprintf("        filesystem-label: %s\n", save)
+		}
+	}
+
+	gi, err := gadget.InfoFromGadgetYaml([]byte(h), nil)
+	c.Assert(err, IsNil)
+	return gi
+}
+
+func (s *validateGadgetTestSuite) TestVolumeRulesConsistencyNoModel(c *C) {
+	ginfo := func(hasSeed bool, label string) *gadget.Info {
```

Comment:
```suggestion
	ginfo := func(hasSeed bool, dataLabel string) *gadget.Info {
```

(this makes it more clear that the label here is the label for the data/writable partition)